### PR TITLE
Added defaultAction for <filter>

### LINKED ISSUE #2151: log filter "action=log" behaving unexpectedly?
When filtering the logging with only one filter with "action=log", this is strange and useless as it won't discard other logevents  

e.g.
```
<logger name="*" writeTo="file">
   <filters>
    <when condition="length('${message}') > 100" action="Log" />
  </filters>
</logger>
``` 

Still *All* logevents will be logged.

Options:

- Or discard others by default when using `action=log` (but not with `action=neutral`?)
- or add an option, "discard-others"
- Or add action logOnly
- add option, default-action (currently _log_)

Related: https://github.com/NLog/NLog/issues/2125
See https://github.com/NLog/NLog/wiki/Filtering-log-messages

### PR DESCRIPTION
**Pull request by 304NotModified**
fixes #2151

### CODE REVIEW

**Comment by NLog-SonarQube** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [LogFactory.cs#L1108](https://github.com/NLog/NLog/blob/15ef65853eb496859f2778a2db8f3c54733cdf3b/src/NLog/LogFactory.cs#L1108): Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1066)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by 304NotModified** [comment]
@snakefoot any opinion on this?

File: src/NLog/LoggerImpl.cs [hunk 1]
```diff
@@ -96,17 +96,20 @@ internal static void Write([NotNull] Type loggerType, TargetWithFilterChain targ
...
 
             IList<Filter> prevFilterChain = null;
+            FilterResult? prevDefaultResult = null;
+
             FilterResult prevFilterResult = FilterResult.Neutral;
             for (var t = targets; t != null; t = t.NextInChain)
             {
-                FilterResult result = ReferenceEquals(prevFilterChain, t.FilterChain) ?
-                    prevFilterResult : GetFilterResult(t.FilterChain, logEvent);
+                FilterResult result = ReferenceEquals(prevFilterChain, t.FilterChain) && Equals(prevDefaultResult, t.DefaultResult) ?
```
Comment (snakefoot) [inline src/NLog/LoggerImpl.cs]: Don't think you need to modify this line. If ReferenceEquals returns true, then everything is identical.

**Comment by snakefoot** [comment]
@304NotModified Besides my comment about removing `prevDefaultResult`, then it looks good.

File: src/NLog/LoggerImpl.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/LoggerImpl.cs]: but `DefaultResult` isn't in `t.FilterChain`? So i could be another one?

Comment (snakefoot) [reply src/NLog/LoggerImpl.cs]: Is it possible to have two targets with the same `FilterChain`-list-reference where the DefaultResult is different? The `FilterChain`-list and `DefaultResult` comes from the same LoggingRule-object (Filters + DefaultFilterResult).

Comment (304NotModified) [reply src/NLog/LoggerImpl.cs]: currently not I think.

OK,thanks, so I can remove 15ef65853eb496859f2778a2db8f3c54733cdf3b from this branch and then it's done.

**Comment by 304NotModified** [comment]
Thanks for checking!

**Comment by NLog-SonarQube** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [LogFactory.cs#L1108](https://github.com/NLog/NLog/blob/7606cb01ec071e97c6384226a72de8bc65e369cc/src/NLog/LogFactory.cs#L1108): Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1066)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by 304NotModified** [comment]
docs added here:

- https://github.com/NLog/NLog/wiki/Filtering-log-messages
- https://github.com/NLog/NLog/wiki/When-filter

**Pull request merged**